src/utilities/user.ts

```
@@ -1,23 +1,5 @@
-import { type EditorState, type UserState } from '@/store/awareness-store';
-import { areSelectionsEqual } from '@/utilities/selection';
-
-function areEditorStatesEqual( state1?: EditorState, state2?: EditorState ): boolean {
-	return Boolean( state1 && state2 && areSelectionsEqual( state1.selection, state2.selection ) );
-}
+import { type UserState } from '@/store/awareness-store';
 
 export function areUserStatesEqual( userState1: UserState, userState2: UserState ): boolean {
-	if ( Object.keys( userState1 ).length !== Object.keys( userState2 ).length ) {
-		return false;
-	}
-
-	return Object.entries( userState1 ).every( ( [ key, value ] ) => {
-		// Update this switch with any non-primitive fields added to UserState.
-		switch ( key ) {
-			case 'editorState':
-				return areEditorStatesEqual( userState1.editorState, userState2.editorState );
-
-			default:
-				return value === userState2[ key as keyof UserState ];
-		}
-	} );
+	return JSON.stringify( userState1 ) === JSON.stringify( userState2 );
```

@chriszarate I changed this whole function to just use `JSON.stringify()`. I did this because it took me a while to track down a bug that caused excessive awareness updates. The cause was `areSelectionsEqual()` not being updated to correctly handle comparison of the new relative positions. Because `areUserStatesEqual()` was always returning `false`, this added a lot of redraws and noise when nothing was actually happening to the document. If the comparison code had been wrong in the other way (incorrectly returning `true`) it would have broken awareness updates for the new data type.

Overall, it does seem like an easy problem to encounter each time we add to user state or make a data type change. It's a minor footgun that can cause subtle performance issues and seems likely to happen again. Instead, this was changed to just use a `JSON.stringify()` comparison. It currently works for all data types in editor state, and unless we add something non-serializable to local state it should work well as a replacement. If we do need to come back to a property-by-property comparison, it should be easy to AI-up that code.

---

> it took me a while to track down a bug that caused excessive awareness updates. The cause was areSelectionsEqual() not being updated to correctly handle comparison of the new relative positions

Sorry about that! I should have coded it more defensively. More ideally it should `throw` an error in the `default` case instead of returning `false`:

https://github.com/Automattic/vip-real-time-collaboration/blob/e49403cc565928d79632c58c72be40e8c71c996a/src/utilities/selection.ts#L96

I think I would prefer this route instead, if possible. `JSON.stringify` has its own performance issues and potential footguns, such as unserializable properties that just get dropped. What do you think?

---

> Sorry about that! I should have coded it more defensively.

This certainly isn't your fault. In this case, I changed the data type of `selection.cursorPosition`, so the non-`default` code still executed, it just started returning `false` where it previously hadn't because a a deeper data type changed. We just need to be really diligent about any changes to editor state, because false negatives will lead to incorrect UI and false positives will lead to subtle performance issues that get worse with more users. Or we should consider a different representation for editor states where the logic is simplified.

For now, I made a few changes:

1. In https://github.com/Automattic/vip-real-time-collaboration/pull/55/commits/7b095dc4de9c67d344ac1b802addfa11bfaa770b, changed comparison back to per-property as before. Note that relative positions are still compared via `JSON.stringify()` as this is [explicitly supported](https://docs.yjs.dev/api/relative-positions). Relative positions look like this: `{"type":{"client":305620264,"clock":1458},"tname":null,"item":{"client":305620264,"clock":1472},"assoc":0}`, so comparing them property-by-property would require careful iteration of the object internals, which feels like unnecessary coupling.
2. Removed a `throw` in https://github.com/Automattic/vip-real-time-collaboration/pull/55/commits/13d37ac89dc1e383fe869146ee96d15c2935f6e2 that I occasionally saw in testing and replaced it with gracefully not drawing a cursor instead.
3. Similar to your suggestion above, I changed the `default` case to log an error in https://github.com/Automattic/vip-real-time-collaboration/pull/55/commits/541a9f74e8d42d672cb7427ddbecc6543c0275ce. I've been trying to remove `throw`s in Awareness code unless it's something fatal. I don't imagine we'll have even more selection types come into play, but if we do and it causes a bug we'll at least have an error to look at.